### DISCUSSION TITLE
MultiheadAttentionBlock: Unusual FFN

### DISCUSSION THREAD

**Discussion by ldv1** [pyg-team/pytorch_geometric#7986] [Q&A] [answered]
Hi,

In the [MultiheadAttentionBlock](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/aggr/utils.py) class, the position-wise feed-forward network consists of a single linear transformation with a ReLU activation:

`out = out + self.lin(out).relu()`

whereas in almost any other implementation it consists of two linear transformations with a ReLU activation in between (based on the paper [Attention Is All You Need](https://arxiv.org/pdf/1706.03762.pdf)):

`out = out + self.lin2(self.lin1(out).relu())`

(and nowadays, GELU is standard)

What is the rational behind this unusual implementation?

**Comment by rusty1s** [comment]
This is mostly taken from the official implementation, see https://github.com/davidbuterez/gnn-neural-readouts/blob/main/code/models/set_transformer_modules.py#L8.

**Comment by ldv1** [comment, marked as answer]
I see.
This seems strange to me but this readout is entirely based on the [Set Transformer: A Framework for Attention-based Permutation-Invariant Neural Networks](https://arxiv.org/abs/1810.00825) by Lee et al.  whose [original implementation](https://github.com/juho-lee/set_transformer)  also uses a single linear layer followed by ReLU.
Thanks for the prompt answer.

**Reply by rusty1s** [reply]
I am happy to change this if you see a strong need to do this.

**Comment by ldv1** [comment]
Related to this topic:

In the implementation you mentioned, we found for the [PMA](https://github.com/davidbuterez/gnn-neural-readouts/blob/main/code/models/set_transformer_modules.py#L57)

```
class PMA(nn.Module):
    def __init__(...)
    ...
    def forward(self, X):
        return self.mab(self.S.repeat(X.size(0), 1, 1), X)
```

but in [PyG](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/aggr/utils.py#L174), we have

```
class PoolingByMultiheadAttention(torch.nn.Module):
    def __init__(...)
    ...
    def forward(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
        x = self.lin(x).relu()
        return self.mab(self.seed.expand(x.size(0), -1, -1), x, y_mask=mask)
```

Why do we have `x = self.lin(x).relu()` that is also not specified in the Set Transformer paper ? (unless I oversaw it) ?
Thanks for the help.

**Reply by rusty1s** [reply]
This is taken from Eq (11) in https://arxiv.org/pdf/1810.00825.pdf. It depends on whether you want to implement the feature transformation within or without the module. The paper does integrate it into `PMA` while the code does not.

**Reply by ldv1** [reply]
You're right! Many thanks!
